src/mainwin.cc

```diff
@@ -2013,6 +2017,164 @@ void MainWindow::csgRender()
 	compileEnded();
 }
 
+
+void MainWindow::action3DPrint()
+{
+	//Keeps track of how many times we've exported and tries to create slightly unique filenames.
+	//Not mission critical, since non-unique file names are fine for the API, just harder to 
+	//differentiate between in customer support later.
+	static unsigned int printCounter=0;
+	
+	setCurrentOutput();
+	PRINT("3D Printing...");
+	
+	unsigned int dim = 3;
+	
+	QUrl partUrl;
+	
+	//Where we hold our temporary stl file name:
+	char tempStlFileName[L_tmpnam];
+	
+	setCurrentOutput();
+
+	//Make sure we can export:
+	if (! canExport(dim))
+	{
+		PRINT("Cannot 3D Print due to errors.");
+		return;
+	}
+	
+	//Ccreate a temporary file name valid on all systems:
+	QString exportFilename=QString(std::tmpnam(tempStlFileName));
```

Comment:
Ok.  I tried your suggestion, but because QTemporary file can only give a file name after the file is open, it ended up creating a lot more code than what I had before.  Here's the diff:

```
index 4fcb4db9..a8775eea 100755
--- a/src/mainwin.cc
+++ b/src/mainwin.cc
@@ -2042,10 +2042,17 @@ void MainWindow::action3DPrint()
  {
    PRINT("Cannot 3D Print due to errors.");
    return;
- }
+     }
  
- //Ccreate a temporary file name valid on all systems:
- QString exportFilename=QString(std::tmpnam(tempStlFileName));
+ //Create a temporary file name valid on all systems:
+ QTemporaryFile exportFile;
+ //Open the file so we can get the name:
+ if ( ! exportFile.open())
+ {
+   PRINT("Could not open temporary file.");
+   return;
+ }
+ QString exportFilename=exportFile.fileName();
  
  //Render the stl to a temporary file:
  exportFileByName(this->root_geom, FileFormat::STL,
@@ -2167,9 +2174,6 @@ bool MainWindow::uploadStlAndGetPartUrl(const QString & exportFilename, const QS
  //Put it in our output partUrl:
  partUrl.setUrl(partUrlStr);
  
-    //Remove the temporary file:
-    remove(exportFilename.toLocal8Bit().constData());
-    
  return true;
 }

```

Reply:
So, I'm not entirely sure that in this specific case, QTemporaryFile is an improvement.  Thoughts?

Reply:
It will also remove the buffer `tempStlFileName` and I think the automatic resource handling is useful. In addition it's not a fair comparison as `std::tmpnam()` *may* return NULL even when given a buffer, so that (probably unlikely) state is not checked.

Reply:
Good point, yeah the automatic resource handling is better than a manual remove().  Agreed.   Thanks!